src/snmalloc/backend/backend.h

```diff
@@ -28,132 +28,64 @@ namespace snmalloc
        */
-      SNMALLOC_FAST_PATH
-      static void set_metaentry(address_t p, size_t size, const Entry& t)
-      {
-        for (address_t a = p; a < p + size; a += MIN_CHUNK_SIZE)
-        {
-          concretePagemap.set(a, t);
-        }
-      }
+      friend class FlatPagemap<MIN_CHUNK_BITS, PageMapEntry, PAL, fixed_range>;
 
-    public:
       /**
-       * Get the metadata associated with a chunk.
-       *
-       * Set template parameter to true if it not an error
-       * to access a location that is not backed by a chunk.
+       * The only constructor that creates newly initialised meta entries.
+       * This is callable only by the back end.  The front end may copy,
+       * query, and update these entries, but it may not create them
+       * directly.  This contract allows the back end to store any arbitrary
+       * metadata in meta entries when they are first constructed.
        */
-      template<bool potentially_out_of_range = false>
-      SNMALLOC_FAST_PATH static const auto& get_metaentry(address_t p)
-      {
-        return concretePagemap.template get<potentially_out_of_range>(p);
-      }
+      SNMALLOC_FAST_PATH
+      PageMapEntry(SlabMetadata* meta, uintptr_t ras)
+      : FrontendMetaEntry<SlabMetadata>(meta, ras)
+      {}
 
       /**
-       * Get the metadata associated with a chunk.
-       *
-       * Set template parameter to true if it not an error
-       * to access a location that is not backed by a chunk.
+       * Copy assignment is used only by the pagemap.
        */
-      template<bool potentially_out_of_range = false>
-      SNMALLOC_FAST_PATH static auto& get_metaentry_mut(address_t p)
-      {
-        return concretePagemap.template get_mut<potentially_out_of_range>(p);
-      }
-
-      static void register_range(address_t p, size_t sz)
+      PageMapEntry& operator=(const PageMapEntry& other)
       {
-        concretePagemap.register_range(p, sz);
+        FrontendMetaEntry<SlabMetadata>::operator=(other);
+        return *this;
       }
 
       /**
-       * Return the bounds of the memory this back-end manages as a pair of
-       * addresses (start then end).  This is available iff this is a
-       * fixed-range Backend.
+       * Default constructor.  This must be callable from the pagemap.
        */
-      template<bool fixed_range_ = fixed_range>
-      static SNMALLOC_FAST_PATH
-        std::enable_if_t<fixed_range_, std::pair<address_t, address_t>>
-        get_bounds()
-      {
-        static_assert(
-          fixed_range_ == fixed_range, "Don't set SFINAE parameter!");
-
-        return concretePagemap.get_bounds();
-      }
-
-      static bool is_initialised()
-      {
-        return concretePagemap.is_initialised();
-      }
+      SNMALLOC_FAST_PATH PageMapEntry() = default;
     };
 
+    using Pagemap = Pagemap<
+      PAL,
+      FlatPagemap<MIN_CHUNK_BITS, PageMapEntry, PAL, fixed_range>,
```

Comment:
Since this type is used in two places, it would make sense to have a (private) `using` declaration to name the type so that these two can't get out of sync.  Something like:
```
using ConcretePagemap = FlatPagemap<MIN_CHUNK_BITS, PageMapEntry, PAL, fixed_range>;
```
And then use `ConcretePagemap` in both places.